File: backend/scripts/generate_token.py
#!/usr/bin/env python3
"""
Generate a JWT API token for frontend authentication.

Usage:
    python3 scripts/generate_token.py [--expires-days DAYS] [--secret SECRET]

Example:
    python3 scripts/generate_token.py --expires-days 365
    python3 scripts/generate_token.py --secret "your-secret-here"
"""
import sys
import os
import argparse
from datetime import datetime, timedelta

# Try to use backend modules if available, otherwise use standalone implementation
try:
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
    from backend.utils.jwt import create_api_token
    from backend.config import settings
    USE_BACKEND_MODULES = True
except ImportError:
    USE_BACKEND_MODULES = False
    try:
        import jwt
    except ImportError:
        print("Error: Either backend dependencies or PyJWT must be installed.")
        print("Install with: pip install PyJWT")
        print("Or run from within the backend Docker container.")
        sys.exit(1)


def create_token_standalone(secret: str, expires_days: int = 365) -> str:
    """Standalone token creation using PyJWT."""
    payload = {
        "type": "api_token",
        "exp": datetime.utcnow() + timedelta(days=expires_days),
        "iat": datetime.utcnow(),
    }
    return jwt.encode(payload, secret, algorithm="HS256")


def main():
    parser = argparse.ArgumentParser(description='Generate JWT API token for frontend')
    parser.add_argument(
        '--expires-days',
        type=int,
        default=365,
        help='Number of days until token expires (default: 365)'
    )
    parser.add_argument(
        '--secret',
        type=str,
        default=None,
        help='JWT secret (if not provided, uses JWT_SECRET env var or backend config)'
    )
    args = parser.parse_args()
    
    # Get JWT secret
    if args.secret:
        jwt_secret = args.secret
    elif USE_BACKEND_MODULES:
        jwt_secret = settings.jwt_secret
        if jwt_secret == "changeme-in-production":
            print("WARNING: JWT_SECRET is set to default value!")
            print("Set JWT_SECRET environment variable or use --secret flag.")
            print("Example: export JWT_SECRET=$(openssl rand -hex 32)")
            sys.exit(1)
    else:
        jwt_secret = os.getenv('JWT_SECRET')
        if not jwt_secret:
            print("Error: JWT_SECRET not set!")
            print("Set JWT_SECRET environment variable or use --secret flag.")
            print("Example: export JWT_SECRET=$(openssl rand -hex 32)")
            sys.exit(1)
    
    try:
        if USE_BACKEND_MODULES:
            token = create_api_token(expires_days=args.expires_days)
        else:
            token = create_token_standalone(jwt_secret, args.expires_days)
        
        print(f"\nAPI Token (expires in {args.expires_days} days):")
        print(f"{token}\n")
        print("Set this as NEXT_PUBLIC_API_TOKEN in your frontend environment variables.")
    except Exception as e:
        print(f"Error generating token: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: backend/scripts/generate_token.js
#!/usr/bin/env node
/**
 * Generate a JWT API token for frontend authentication.
 * 
 * Usage:
 *   node scripts/generate_token.js [--secret SECRET] [--expires-days DAYS]
 * 
 * Example:
 *   node scripts/generate_token.js --secret "your-secret-here"
 */

const crypto = require('crypto');

function base64UrlEncode(str) {
  return Buffer.from(str).toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

function createJWT(secret, expiresDays = 365) {
  const header = {
    alg: 'HS256',
    typ: 'JWT'
  };

  const now = Math.floor(Date.now() / 1000);
  const payload = {
    type: 'api_token',
    exp: now + (expiresDays * 24 * 60 * 60),
    iat: now
  };

  const encodedHeader = base64UrlEncode(JSON.stringify(header));
  const encodedPayload = base64UrlEncode(JSON.stringify(payload));

  const signature = crypto
    .createHmac('sha256', secret)
    .update(`${encodedHeader}.${encodedPayload}`)
    .digest('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');

  return `${encodedHeader}.${encodedPayload}.${signature}`;
}

// Parse command line arguments
const args = process.argv.slice(2);
let secret = null;
let expiresDays = 365;

for (let i = 0; i < args.length; i++) {
  if (args[i] === '--secret' && args[i + 1]) {
    secret = args[i + 1];
    i++;
  } else if (args[i] === '--expires-days' && args[i + 1]) {
    expiresDays = parseInt(args[i + 1], 10);
    i++;
  }
}

if (!secret) {
  secret = process.env.JWT_SECRET;
}

if (!secret) {
  console.error('Error: JWT_SECRET not set!');
  console.error('Set JWT_SECRET environment variable or use --secret flag.');
  console.error('Example: export JWT_SECRET=$(openssl rand -hex 32)');
  process.exit(1);
}

try {
  const token = createJWT(secret, expiresDays);
  console.log(`\nAPI Token (expires in ${expiresDays} days):`);
  console.log(token);
  console.log('\nSet this as NEXT_PUBLIC_API_TOKEN in your frontend environment variables.\n');
} catch (error) {
  console.error(`Error generating token: ${error.message}`);
  process.exit(1);
}

